_posts/2024-08-22-boost-vector-search-with-css.md

```diff
@@ -0,0 +1,574 @@
+
+```json
+PUT _cluster/settings
+{
+   "persistent": {
+      "search.concurrent_segment_search.enabled": true
+   }
+}
+```
+
+To achieve concurrent segment searches, OpenSearch divides the segments within each shard into multiple slices, with each slice processed in parallel on a separate thread. The number of slices determines the degree of parallelism that OpenSearch can provide. You can either use Lucene's default slicing mechanism or set the maximum slice count manually. For detailed instructions on updating the slice count, see [Slicing mechanisms](https://opensearch.org/docs/latest/search-plugins/concurrent-segment-search/#slicing-mechanisms).
+
+## Performance results
+
+We performed our tests on an [OpenSearch 2.15](https://opensearch.org/versions/opensearch-2-15-0.html) cluster using the OpenSearch Benchmark [vector search workload](https://github.com/opensearch-project/opensearch-benchmark-workloads/tree/main/vectorsearch). We used the Cohere dataset with two different configurations to evaluate the performance improvements of vector search queries when running the workload with concurrent segment search disabled, enabled with default settings, and enabled with different max slice counts.
+
+### Cluster setup
+
+- 3 data nodes (r5.4xlarge: 128 GB RAM, 16 vCPUs, 250 GB disk space)
+- 3 cluster manager nodes (r5.xlarge: 32 GB RAM, 4 vCPUs, 50 GB disk space)
+- 1 OpenSearch workload client (c5.4xlarge: 32 GB RAM, 16 vCPUs)
+- 1 and 4 search clients
+- `index_searcher` thread pool size: 32
+
+#### Index settings
+
+|`m`	|`ef_construction`	|`ef_search`	|Number of shards	|Replica count	|Space type	|
+|---	|---	|---	|---	|---	|---	|
+|16	|100	|100	|6	|1	|inner product	|
+
+#### Configuration
+
+|Dimension	|Vector count	| Search query count	|Refresh interval	|
+|:--- | :--- | :--- | :--- |
+|768	| 10M	|10K	|1s (default)	|
+
+### Service time comparison
+
+We conducted the following experiments:
+
+1. [Concurrent search disabled](#experiment-1-concurrent-search-disabled)
+1. Concurrent search enabled:
+   - [Max slice count = 0 (default)](#experiment-2-concurrent-search-enabled-max-slice-count--0-default)
+   - [Max slice count = 2](#experiment-3-concurrent-search-enabled-max-slice-count--2)
+   - [Max slice count = 4](#experiment-4-concurrent-search-enabled-max-slice-count--4)
+   - [Max slice count = 8](#experiment-5-concurrent-search-enabled-max-slice-count--8)
+
+The following sections present the results of these experiments.
+
+#### Experiment 1: Concurrent search disabled
+
+<table border="1">
+  <thead>
+    <tr>
+      <th>k-NN engine</th>
+      <th>Segment count</th>
+      <th>Num search clients</th>
+      <th colspan="3">Service time (ms)</th>
+      <th>Max CPU %</th>
+      <th>% JVM heap used</th>
+      <th>Recall</th>
+    </tr>
+    <tr>
+      <th></th>
+      <th></th>
+      <th></th>
+      <th>p50</th>
```

Comment:
can we include something like `P50(in ms) ` in other places wherever we are displaying service time number?

Reply:
@VijayanB - do you still have edits to make? I was about to merge this one (cc @pajuric )

Reply:
@VijayanB - Fanit is going to make the edit - but you need to let her know exactly what you want this to be, please.

Reply:
@VijayanB Actually, if you look at the rendered form, you don't need the (ms) because the p50, p90, p99 are under a common heading "Service time (ms)"
![Screenshot 2024-08-27 at 5 31 32 PM](https://github.com/user-attachments/assets/535aaadf-e4e6-4d1e-8427-98183dfb3a67)